core/src/banking_stage.rs

```diff
@@ -1077,6 +1026,287 @@ impl BankingStage {
+                        )
+                        .map(|transaction| (transaction, packet_index))
+                    } else {
+                        None
+                    }
+                })
+                .unzip(),
+            "sanitize_packet",
+        );
+
+        // report metrics
+        let packet_conversion_us = packet_conversion_time.as_us();
+        slot_metrics_tracker.increment_transactions_from_packets_us(packet_conversion_us);
+        banking_stage_stats
+            .packet_conversion_elapsed
+            .fetch_add(packet_conversion_us, Ordering::Relaxed);
+        inc_new_counter_info!("banking_stage-packet_conversion", 1);
+        let unsanitized_packets_filtered_count =
+            packets_to_process.len().saturating_sub(transactions.len());
+        inc_new_counter_info!(
+            "banking_stage-dropped_tx_before_forwarding",
+            unsanitized_packets_filtered_count
+        );
+
+        (transactions, transaction_to_packet_indexes)
+    }
+
+    /// Checks sanitized transactions against bank, returns valid transaction indexes
+    fn filter_invalid_transactions(
+        transactions: &[SanitizedTransaction],
+        bank: &Arc<Bank>,
+        banking_stage_stats: &BankingStageStats,
+    ) -> Vec<usize> {
+        let (results, filter_invalid_transactions_time) = measure!(
+            {
+                let filter = vec![Ok(()); transactions.len()];
+                Self::bank_check_transactions(bank, transactions, &filter)
+            },
+            "filter_invalid_transactions"
+        );
+
+        // report metrics
+        let filter_invalid_transactions_us = filter_invalid_transactions_time.as_us();
+        banking_stage_stats
+            .filter_pending_packets_elapsed
+            .fetch_add(filter_invalid_transactions_us, Ordering::Relaxed);
+        let filtered_out_transactions_count = transactions.len().saturating_sub(results.len());
+        inc_new_counter_info!(
+            "banking_stage-dropped_tx_before_forwarding",
+            filtered_out_transactions_count
+        );
+
+        results
+            .iter()
+            .enumerate()
+            .filter_map(
+                |(tx_index, (result, _))| if result.is_ok() { Some(tx_index) } else { None },
+            )
+            .collect_vec()
+    }
+
+    fn prepare_filtered_packet_indexes(
+        transaction_to_packet_indexes: &[usize],
+        retained_transaction_indexes: &[usize],
+    ) -> Vec<usize> {
+        retained_transaction_indexes
+            .iter()
+            .map(|tx_index| transaction_to_packet_indexes[*tx_index])
+            .collect_vec()
+    }
+
+    fn collect_retained_packets(
+        buffered_packet_batches: &mut UnprocessedPacketBatches,
+        packets_to_process: &[Rc<ImmutableDeserializedPacket>],
+        retained_packet_indexes: &[usize],
+    ) -> Vec<Rc<ImmutableDeserializedPacket>> {
+        // Remove the non-retainable packets
+        Self::filter_processed_packets(
```

Comment:
I think we can rip this logic out into a separate method, I remember it being used in ```consume_buffered_packets``` as well

Reply:
✅